I18n weather messages and days (fixes #24)

janus2 opened the pull request.

humdingerb commented:
The weather conditions should be sentence-cased, e.g. "Mostly cloudy".

Not sure if it's worth the effort, but you may be able to get the localized day abbreviations from the locale kit. haiku/src/kits/shared/CalendarView.cpp may give some clues.
This would make the replicant's GUI almost locale-neutral, if this is done as well:

If there isn't a free i18n aware weather service around, a crutchy solution could be to have another BTextControl in the "Change location..." window to "Display as:", where the user enters how the location is displayed.
That would even allow for "Home", "@Granny's" etc. :)

janus2 commented:
You are right on the sentence-case but I used the string returned by the service...
For the days of the week it is a lot of work those are the string present in the documentation of the service, it is more code to convert the date with the locale kit that translate five strings.

humdingerb commented:
> You are right on the sentence-case but I used the string returned by the service...

I'm confused. The service returns an int32 condition code. The _GetWeatherMessage() returns our translated string. Do you agree on the sentence casing and modify aour PR?

> For the days of the week it is a lot of work those are the string present in the documentation of the service, it is more code to convert the date with the locale kit that translate five strings.

Yes, but OTOH, if those day abbrev. wouldn't have to be explicitely translated, the replicant may become completely locale-agnostic. I don't really know how much code is involved. Doesn't something like this work:
```
BLanguage language("en");
BFormattingConventions formatting("en_US");
BDateFormat format(language, formatting);
BString buffer;
status_t result = format.GetDayName(0, buffer, B_SHORT_DATE_FORMAT);
```
Just an idea, I totally trust your decision... :)

janus2 commented:
I associated the code to the string returned by the service we can modify the string if we want. This is how the app behaves now you can see the screenshot on the readme of this 
 repository "Mostly Cloudy".

humdingerb commented:
Sentence casing would fit in better with the rest of the system.

janus2 commented:
Ok I will change the strings and I'll see the translation of the day of the week with the locale

humdingerb commented:
You're a prince! :)

janus2 commented:
![thesentencecaseisright](https://user-images.githubusercontent.com/10423645/32993997-16d3dbd6-cd61-11e7-9768-c79702ab0ac7.png)

janus2 commented:
DE
![de](https://user-images.githubusercontent.com/10423645/32994432-a0d5dd7e-cd67-11e7-89f1-430002a09a06.png)
US
![us](https://user-images.githubusercontent.com/10423645/32994433-a502b336-cd67-11e7-9f2f-256885f08cb5.png)
It is very strange...

humdingerb commented:
You mean the missing "Menlo"? Otherwise, the abbreviated day names look OK, no?

humdingerb commented:
oh, never mind, you changed location to Park, Mt... :)

janus2 commented:
In US the abbreviations are Sun, Mon, Tue, Wed from the server Su, Mo, Tu from locale
The DE version is OK?

janus2 commented:
LOL for the 3 chars name the parameter is B_LONG_DATE_FORMAT...

humdingerb commented:
heh. I wonder what's the constant is for a full day name, like "Monday"...

We have enough space to use 3 letter abbreviations. Otherwise, 2 would be OK too, like in Dekbar's calendar.

janus2 commented:
![haiku](https://user-images.githubusercontent.com/10423645/32995057-bdffe4fa-cd6f-11e7-87c5-ec98cf275a15.png)

humdingerb commented:
Great. If you add a "Fixes #26" to the commit message, it'll close the issue automatically.

Also, if I see this correctly, we'll now try for three times to connect and only fail then. Between each try there's a 1 minute delay. Right? What do you think of trying every 5 seconds instead and try 30 times or so. Otherwise, the replicant will still say "Connection error" for a minute, when the reconnect would probably have succeeded 5 seconds after bootup, after the wifi hooked up.

janus2 commented:
The time out is around 1 minute so the app is in loading... for 1 minute if there is an error I set the reload to 1 min, 2 min, 3 min after that it goes to the max time for reload 4hr... if we receive an answer is the server that sets how often to reload the data. If the problem is on the server side we overload the server with all the haiku users :-)

humdingerb commented:
I see. I guess there's no easy way to check if the network is up and wait for that before sending the first request?

janus2 commented:
Maybe in a future version, we must start a  monitor on the network status and verify the transition on the different interfaces. It is definitely the best solution.

janus2 commented:
Cool. Cool, cool, cool.

humdingerb commented:
That looks quite elegant (though I haven't followed the code in every detail)... :)
The replicant still doesn't connect as soon as the network comes up, but about a minute later. Can it be made to wait with sending the first service-request until the network came up? Then the Replicant would display the info as soon as it's possible.

janus2 commented:
It  already does that it is very strange I use a VM and the connection start as soon as i set the link to connected.
Do you have only one weather app on your machine? This is the behavior of e62ee96 not 68f2f85.
Do you see "connecting..." or "connection error"?

janus2 commented:
Maybe on the real hardware when the notification arrive the connection is not completely ready.
Very strange.

![disconnected](https://user-images.githubusercontent.com/10423645/33030890-d6e90fe8-ce1c-11e7-9153-73f276a0a379.png)
![connected](https://user-images.githubusercontent.com/10423645/33030892-d891b354-ce1c-11e7-80e2-2da1c3eef86d.png)

humdingerb commented:
RIght after bootup, before the wifi connection is up, I see a "Connecting..." then after a second it turns to "Connection error". And a second after that, the wifi connection is established.
Maybe you can simulate the wifi delay by unplugging the LAN cable (if you're using LAN) and only re-plug it 5 seconds after bootup.

janus2 commented:
it works, if you disable the wifi force e reload with the button and after that you enable the wifi again it works?

humdingerb commented:
Since my wlan is flaky when it comes to un/reconnecting, I used LAN for my tests:

Scenario 1:
When the system is running and Weather is working (all done with the replicant on the Desktop).

- un-plug LAN
- click icon
- status turns into "Loading..."
 -> re-plug LAN : works as expected

- un-plug LAN
- click icon
- status turns into "Loading..."
- after a minute the status turns into "Connecting..."
 -> re-plug LAN : works as expected


Scenario 2:
When booting with un-plugged LAN (all done with the replicant on the Desktop).

- status shows "Connecting..."
 -> re-plug LAN : "Connection error"
- click icon
 -> status turns to "Loading..." then shows weather condition

So, scenario 2 isn't quite correct, re-plugging shouldn't result in an error.

Maybe the status should be "No network" while there's no network connection and only turn to "Connection..." for the server request. When does it say "Loading..."?

janus2 commented:
There is some problem with the real hardware after re-plug LAN the status should be loading... without the click on the icon. 
I have added a delay on the reconnection after the notification that the network is ok... how much time do you think is enough? without real hardware i cannot test this cases.
I have changed "connecting..." to "no connection" buy your "No network" is better.

janus2 commented:
I have added 5 sec before a reconnection. THIS MUST WORK!!! I hope...

humdingerb commented:
Cured! :)
I doe work with my delayed wifi connection. And even when booting with an unplugged LAN and waiting roughly 30 seconds to plug in the cable, it still picks up on it and loads the data.
Well done!

I have to go for today, but I'll merge your PR tomorrow, so if you'd like to do any polishing or so, you've got a bit of time. :)
Thanks a lot!

janus2 commented:
I think it needs a lot of testing on real hardware...
